File: utils/Paginator.py
import discord
import asyncio
import re


class Paginator:
    """
    A class that implements pagination to the bot.
    Note: pages are based of a 1 based index not 0 based index.

    Parameters
    ------------
    ctx: Context
        The command context.
    pages: List[discord.Embed()]
        A list of embeds to paginate.
    color: int
        The color of the embed.
        This defaults to white (0xffffff)
    """

    def __init__(self, ctx, pages=[], color=0xffffff, footer=True):
        self.ctx = ctx
        self.bot = ctx.bot
        self.pages = pages
        self.color = color
        self.footer = footer
        self.reactions = ["⏪", "◀", "⏹", "▶", "⏩", "❔"]
        self.message = None
        self.enabled = False
        self.current = 0

    async def react(self):
        for reaction in self.reactions:
            if len(self.pages) < 2 and reaction in "⏪⏩":
                continue
            try:
                await self.message.add_reaction(reaction)
            except discord.Forbidden:
                pass

    def is_valid_page(self, page: int):
        if page == -1 or page == len(self.pages):
            return False
        return True

    async def switch_page(self, page: int):
        if not self.is_valid_page(page):
            return  # Wait for a valid page
        self.current = page

        if self.enabled:
            to_switch_to = self.pages[page]
            to_switch_to.color = self.color
            if self.footer:
                to_switch_to.set_footer(text=f"Showing page {page + 1} out of {len(self.pages)}")
            await self.message.edit(embed=to_switch_to)
        else:
            self.enabled = True
            self.pages[0].color = self.color
            if self.footer:
                self.pages[0].set_footer(text=f"Showing page {page + 1} out of {len(self.pages)}")
            self.message = await self.ctx.send(embed=self.pages[0])
            await self.react()

    def check(self, r, u):
        return r.message.id == self.message.id and u.id == self.ctx.author.id and r.emoji in self.reactions

    async def start(self):
        if not self.enabled:
            await self.switch_page(0)
        while self.enabled:
            reaction, user = await self.bot.wait_for("reaction_add", check=self.check)
            try:
                await self.message.remove_reaction(reaction.emoji, user)
            except discord.Forbidden:
                continue
            else:
                if reaction.emoji == "⏪":
                    await self.first_page()
                if reaction.emoji == "◀":
                    await self.backward()
                if reaction.emoji == "⏹":
                    await self.end()
                if reaction.emoji == "▶":
                    await self.forward()
                if reaction.emoji == "⏩":
                    await self.last_page()
                if reaction.emoji == "❔":
                    await self.show_help()

    async def first_page(self):
        await self.switch_page(0)

    async def last_page(self):
        await self.switch_page(len(self.pages) - 1)

    async def forward(self):
        await self.switch_page(self.current + 1)

    async def backward(self):
        await self.switch_page(self.current - 1)

    async def show_help(self):
        help_list = [
            "What do these buttons do?\n",
            "⏪ - Takes you to the first page.",
            "◀ - Takes you back a page.",
            "⏹ - Stops the paginator.",
            "▶ - Takes you forward a page.",
            "⏩ - Takes you to the last page.",
            "❔ - Shows this page."
        ]
        help_embed = discord.Embed(color=self.color)
        help_embed.add_field(name="Paginator Help", value="\n".join(help_list))
        help_embed.set_footer(text=f"Reverting back to the last page in 10 seconds.")
        await self.message.edit(embed=help_embed)
        await asyncio.sleep(10)
        await self.switch_page(self.current)

    async def end(self):
        self.enabled = False
        try:
            await self.message.delete()
        except discord.NotFound:
            pass

File: framework/Paginator.js
const { MessageEmbed } = require("discord.js");

class Paginator {

    constructor(msg, template = new MessageEmbed(), pages = [], useTemplateFooter = false) {
        this.msg = msg;
        this.template = template;
        this.collectors = {
            reaction: null,
            message: null
        };
        this.sentMsg = null;
        this.reactor = msg.author;
        this.pages = pages;
        this.currentPage = 0;
        this.enabled = false;
        this.emotes = ["⏪", "⬅", "➡", "⏩", "⏸", "🔢"];
        this.useTemplateFooter = useTemplateFooter;
    }

    async start() {
        if (!this.enabled) await this.switchPage(0);
        this.collectors.reaction = this.sentMsg.createReactionCollector((reaction, user) => this.emotes.includes(reaction.emoji.name) && user.id === this.reactor.id && reaction.users.remove(user.id).catch(() => null),{ time: 864e5 });
        this.collectors.reaction.on("collect", async r => {
            switch (r.emoji.name) {
                case "⏪": return await this.firstPage();
                case "⬅": return await this.backward();
                case "➡": return await this.forward();
                case "⏩": return await this.lastPage();
                case "⏸": return await this.end();
                case "🔢": return await this.userInputPageSwitch();
            }
        });
        this.collectors.reaction.on("end", () => this.end());
    }

    addPage(template) {
        if (template instanceof MessageEmbed) return this.pages.push(template);
        return this.pages.push(template.call(null, new MessageEmbed()));
    }

    async switchPage(pageNum) {
        if (this.enabled) {
            if (pageNum <= -1) return; // Negative index
            if (pageNum >= this.pages.length) return; // Past or equal to the total pages.
            this.currentPage = pageNum;
            return this.sentMsg.edit(this.generateEmbed(this.pages[this.currentPage]));
        } else {
            this.enabled = true;
            this.sentMsg = await this.msg.channel.send(this.generateEmbed(this.pages[0]));
            for (const emote of this.emotes) {
                if (["⏪", "⏩", "🔢"].includes(emote) && this.pages.length < 2) continue;
                await this.sentMsg.react(emote).catch(() => null);
            }
        }
    }

    async forward() {
        return await this.switchPage(this.currentPage + 1);
    }

    async backward() {
        return await this.switchPage(this.currentPage - 1);
    }

    async lastPage() {
        return await this.switchPage(this.pages.length - 1);
    }

    async firstPage() {
        return await this.switchPage(0);
    }

    async userInputPageSwitch() {
        const tm = await this.msg.channel.send("What page would you like to go to?\n**NOTE: This times out in 20 seconds, you can reply with `cancel`, `stop` to stop this selection.**");
        this.collectors.message = this.msg.channel.createMessageCollector(m => m.author.id === this.reactor.id, { time: 20000, errors: ["time"] });
        this.collectors.message.on("collect", async m => {
            if (m.content.search(/(cancel|quit)/i) > -1) {
                await tm.delete({ timeout: 2500 });
                return this.collectors.message.stop();
            }
            if (!this.pages[parseInt(m.content) - 1]) {
                const temp = await this.msg.channel.send(`Invalid page provided \`[${isNaN(m.content) ? m.content : parseInt(m.content)}/${this.pages.length}\`]`)
                await temp.delete({ timeout: 2500 });
            } else {
                this.switchPage(parseInt(m.content) - 1);
            }
            await tm.delete();
            this.collectors.message.stop();
        });
        this.collectors.message.on("end", async ({ size }) => {
            if (size < 1) {
                const t = await this.msg.channel.send("The selection timed out!");
                return await t.delete({ timeout: 2500 });
            }
        });
    }

    async end() {
        if (this.collectors.message && !this.collectors.message.ended) this.collectors.message.stop();
        await this.sentMsg.delete().catch(() => null);
    }

    generateEmbed(embed) {
        if (this.template) {
            if (this.template.author) {
                const { name, iconURL, url } = this.template.author;
                embed.setAuthor(name, iconURL, url);
            }
            if (this.template.description) embed.setDescription(this.template.description);
            if (this.template.color) embed.setColor(this.template.color);
            if (this.template.fields.length > 0 && embed.fields.length > 0) embed.fields = [...this.template.fields, ...embed.fields];
        }
        embed.setTimestamp();
        embed.setFooter(!(this.useTemplateFooter && this.template.footer) ?
            `Showing page [${this.currentPage + 1}/${this.pages.length}]` :
            this.template.footer.text,
        );
        return embed;
    }

};

module.exports = Paginator;